src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -55966,5 +55966,227 @@ static void M(object x)
+        try
+        {
+        }
+        finally
+        {
+            s = null;
+        }
+
+        _ = s.Length; // warning
+    }
+}
+";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (14,13): warning CS8602: Possible dereference of a null reference.
+                //         _ = s.Length; // warning
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(14, 13)
+                );
+        }
+
+        [Fact, WorkItem(30561, "https://github.com/dotnet/roslyn/issues/30561")]
+        public void SetNullableStateInFinally_02()
+        {
+            var source =
+@"public static class Program
+{
+    public static int Main()
+    {
+        string? s = string.Empty;
+        try
+        {
+            s = null;
+            MayThrow();
+            s = string.Empty;
+        }
+        catch (System.Exception)
+        {
+        }
+
+        return s.Length; // warning: possibly null
+    }
+    static void MayThrow()
+    {
+        throw null;
+    }
+}
+";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (16,16): warning CS8602: Possible dereference of a null reference.
+                //         return s.Length; // warning: possibly null
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(16, 16)
+                );
+        }
+
+        [Fact, WorkItem(30561, "https://github.com/dotnet/roslyn/issues/30561")]
+        public void SetNullableStateInFinally_03()
+        {
+            var source =
+@"public static class Program
+{
+    public static int Main()
+    {
+        string? s = string.Empty;
+        try
+        {
+            s = null;
+            MayThrow();
+            s = string.Empty;
+        }
+        catch (System.Exception)
+        {
+            return s.Length; // warning: possibly null
+        }
+
+        return s.Length;
+    }
+    static void MayThrow()
+    {
+        throw null;
+    }
+}
+";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (14,20): warning CS8602: Possible dereference of a null reference.
+                //             return s.Length; // warning: possibly null
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(14, 20)
+                );
+        }
+
+        [Fact, WorkItem(30561, "https://github.com/dotnet/roslyn/issues/30561")]
+        public void SetNullableStateInFinally_04()
+        {
+            var source =
+@"public static class Program
+{
+    public static int Main()
+    {
+        string? s = string.Empty;
+        try
+        {
+            s = null;
+            MayThrow();
+            s = string.Empty;
+        }
+        catch (System.Exception)
```

Comment:
>catch (System.Exception) [](start = 8, length = 24)

test ideas: 
Consider testing with `catch` that catches some other exception type? I think this will not affect the analysis, but when I saw the tests, I assumed that `System.Exception` was handled specially (as a catch all).

Also, consider some tests with multiple `catch` blocks.

Reply:
`System.Exception` is not handled specially.

Adding a test for multiple catch blocks.

Not adding nested scenarios because it is all compositional. There is no special handling of nesting for try blocks. Trying to test nesting of constructs would cause an explosion of the tests.

---
In reply to: [232864656](https://github.com/dotnet/roslyn/pull/31082#discussion_r232864656) [](ancestors = 232864656)

Reply:
Actually, I will add some nested scenarios.  There is nesting-specific state handling code.

---
In reply to: [233151144](https://github.com/dotnet/roslyn/pull/31082#discussion_r233151144) [](ancestors = 233151144,232864656)